### PR TITLE
Corrected Dijkstra and added some test cases.

### PR DESCRIPTION
**Pull request by cherrygot-personal**

### CODE REVIEW

File: src/iterators/dijkstra.rs [hunk 1]
```diff
@@ -71,6 +72,12 @@ impl<'a, T> Dijkstra<'a, T>
         if graph.fetch(src).is_none() {
             return Err(GraphErr::NoSuchVertex);
         }
+        
+        for edge in graph.edges() {
+            if graph.weight(edge.1, edge.0).unwrap() < 0.0 {
```
Comment (octavonce) [inline src/iterators/dijkstra.rs]: If the edge weight is not set this will panic by unwrapping. Can you please put a case handling an unset weight? When an edge's weight is unset it means that it is 0.

**Review by octavonce — CHANGES_REQUESTED** [review]
Hi there and good job on opening another PR! 

I left you some pointers in the review. It also seems that the Travis CI check didn't work for the latest commit. 

Could you please investigate that prior to merging? It could just be a Travis issue but let's double check to be sure. I had a similar problem yesterday so it might just require another push.

Cheers!

File: src/iterators/dijkstra.rs (hunk 1 above)
Comment (cherrygot-personal) [reply src/iterators/dijkstra.rs]: Oh, I totally forgot that. Will fix it soon.

Comment (cherrygot-personal) [reply src/iterators/dijkstra.rs]: I've corrected the case. Turns out, I haven't considered the same scenario in one more place. Thanks for highlighting it out.

**Comment by octavonce** [comment]
@CherryGot I believe this looks good now. Could you please also rebase the branch with master and then we can proceed with merging?

Cheers!

File: src/iterators/dijkstra.rs [hunk 2]
```diff
@@ -154,13 +169,18 @@ impl<'a, T> Dijkstra<'a, T>
             
             for neighbour in self.iterable.out_neighbors(&vert_meta.id) {
                 if !visited.contains(&neighbour) {
-                    let alt_dist = self.distances.get(&vert_meta.id).unwrap() + self.iterable.weight(&vert_meta.id, &neighbour).unwrap();
+                    let mut alt_dist = self.distances.get(&vert_meta.id).unwrap().clone(); 
+                    
+                    match self.iterable.weight(&vert_meta.id, &neighbour) {
+                        Some(w) => alt_dist += w,
```
Comment (fogti) [inline src/iterators/dijkstra.rs]: This `match` statement could be replaced with a more idiomatic `if let` statement:

```rust
if let Some(w) = self.iterable.weight(&vert_meta.id, &neighbour) {
    alt_dist += w;
}
```

File: src/iterators/dijkstra.rs [hunk 3]
```diff
@@ -71,6 +72,17 @@ impl<'a, T> Dijkstra<'a, T>
         if graph.fetch(src).is_none() {
             return Err(GraphErr::NoSuchVertex);
         }
+        
+        for edge in graph.edges() {
+            match graph.weight(edge.1, edge.0) {
+                Some(w) => {
```
Comment (fogti) [inline src/iterators/dijkstra.rs]: This match statement could be replaced with the following, more idiomatic code:
```rust
if let Some(w) = graph.weight(edge.1, edge.0) {
    if w < 0.0 {
        return Err(GraphErr::InvalidWeight);
    }
}
```

**Review by fogti — CHANGES_REQUESTED** [review]
There are some places in this PR where `match/Some` statements should be replaced with `if let` statements.

File: src/iterators/dijkstra.rs [hunk 4]
```diff
@@ -162,13 +169,18 @@ impl<'a, T> Dijkstra<'a, T>
...
+                        Some(w) => alt_dist += w,
+                        None => {}
+                    }
                     
                     if alt_dist < *self.distances.get(&neighbour).unwrap() {
                         self.distances.insert(*neighbour, alt_dist);
                         self.previous.insert(*neighbour, Some(vert_meta.id));
                         
-                        vertex_pq.push(  VertexMeta {
+                        vertex_pq.push( VertexMeta {
```
Comment (fogti) [inline src/iterators/dijkstra.rs]: why did you remove one leading space and not both?

**Comment by octavonce** [comment]
@CherryGot Do you have an update on this? How soon can we merge?

**Comment by cherrygot-personal** [comment]
@octavonce I'm really sorry for the delay, but my university exams are going on. Will rebase the branch in a day or two.

File: src/iterators/dijkstra.rs (hunk 2 above)
Comment (cherrygot-personal) [reply src/iterators/dijkstra.rs]: Yeah, that's a nice way to write it. Will update in the next PR.

**Comment by octavonce** [comment]
@CherryGot Ok, no worries! Just wanted to know the timeline on this.

**Comment by cherrygot-personal** [comment]
As I updated my fork from purpleprotocol/graphlib, I am unable to test or build. It says something like this,
```Updating crates.io index
error: failed to load source for a dependency on `dot`

Caused by:
  Unable to update registry `https://github.com/rust-lang/crates.io-index`

Caused by:
  failed to fetch `https://github.com/rust-lang/crates.io-index`

Caused by:
  SSL error: unknown error; class=Ssl (16)
```

I tried changing the version of "dot" from cargo.toml but no luck. I don't know, I may be doing something stupid. 
The Travis also failed, stating
```The command "cargo test --verbose --all --features "dot"" exited with 101.```

File: src/iterators/dijkstra.rs [hunk 5]
```diff
@@ -1,6 +1,8 @@
 // Copyright 2019 Chakrapani Gautam
 
-use crate::graph::{Graph, GraphErr};
+use crate::graph::{ Graph, GraphErr };
+use crate::edge::Edge;
+use crate::vertex_id::VertexId;
```
Comment (octavonce) [inline src/iterators/dijkstra.rs:5]: `VertexId` is imported twice here.

**Review by octavonce — COMMENTED** [review]
@CherryGot The stacktrace you provided shows that it is an SSL error, it probably happens due to your internet connection as it works elsewhere. However as I look in the CI stacktrace it shows this error:

```
error[E0252]: the name `VertexId` is defined multiple times
 --> src/iterators/dijkstra.rs:7:5
  |
5 | use crate::vertex_id::VertexId;
  |     -------------------------- previous import of the type `VertexId` here
6 | use crate::iterators::vertices::VertexIter;
7 | use crate::vertex_id::VertexId;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ `VertexId` reimported here
  |
  = note: `VertexId` must be defined only once in the type namespace of this module
```

I'm going to merge this and fix these errors.

**Comment by octavonce** [comment]
Also, thank you @zserik for providing your review!

**Review by fogti — COMMENTED** [review]
looks good, thanks.

File: src/iterators/dijkstra.rs (hunk 5 above)
Comment (cherrygot-personal) [reply src/iterators/dijkstra.rs:5]: Oops, that was dumb!
